industrial_taxonomy/pipeline/glass_house/flow.py

```diff
@@ -0,0 +1,196 @@
+@project(name="industrial_taxonomy")
+class JacchammerFlow(FlowSpec):
+    """Fuzzy match Companies House and Glass AI names using Jacchammer.
+
+    Attributes:
+        run_id: The origin run ID of the flow
+        names_x: Companies House company names
+        names_y: Glass AI organisation names
+        clean_names_x: Pre-processed Companies House company names
+        clean_names_y: Pre-processed Glass AI organisation names
+        full_top_matches: Full match results
+    """
+
+    test_mode = Parameter(
+        "test-mode",
+        help="Whether to run in test mode (on a small subset of data)",
+        type=bool,
+        default=lambda _: not current.is_production,
+    )
+    should_clean_names = Parameter(
+        "clean-names",
+        help="Whether to clean names using Jacchammer default `preproc_names`",
+        type=bool,
+        default=False,
+    )
+
+    names_x: "DataFrame"
+    names_y: "DataFrame"
+    clean_names_x: "DataFrame"
+    clean_names_y: "DataFrame"
+    run_id: int
+    full_top_matches: List[FullMatchResult]  # TODO: py37 error
+
+    @step
+    def start(self):
+        """Load raw data"""
+        from industrial_taxonomy.getters.companies_house import (
+            get_name as get_companies_house_names,
+        )
+        from industrial_taxonomy.getters.glass import (
+            get_organisation as get_glass_organisation,
+        )
+
+        self.run_id = current.origin_run_id or current.run_id
+
+        nrows = 10_000 if (self.test_mode and not current.is_production) else None
+
+        self.names_x = get_companies_house_names().name.head(nrows)
+        self.names_y = get_glass_organisation().name.head(nrows)
+
+        self.next(self.clean_names)
+
+    @pip(path="requirements.txt")
+    @step
+    def clean_names(self):
+        """Pre-process names if `self.should_clean_names`."""
+        from jacc_hammer.name_clean import preproc_names
+
+        self.clean_names_x = (
+            preproc_names(self.names_x) if self.should_clean_names else self.names_x
+        )
+        self.clean_names_y = (
+            preproc_names(self.names_y) if self.should_clean_names else self.names_y
+        )
+
+        self.next(self.match)
+
+    @resources(memory=64_000)
+    @pip(path="requirements.txt")
+    @step
+    def match(self):
+        """The core fuzzy matching algorithm"""
+        from itertools import starmap as starmap_
+        from pathlib import Path
+        from tempfile import TemporaryDirectory
+
+        from jacc_hammer.fuzzy_hash import (
+            Cos_config,
+            Fuzzy_config,
+            match_names_stream,
+        )
+        from pandas import DataFrame
+        from toolz.curried import curry, pipe
+        from toolz.curried import map as tmap
+
+        from utils import optimise_chunk
+
+        starmap = curry(starmap_)
+
+        def persist_chunk(chunk_id: int, chunk: DataFrame) -> int:
```

Comment:
Is there a reason the datastore can't be used for batch processes?

Reply:
It can be and should be by default but sadly it doesn't work for us here.

The full output of the matching is a MASSIVE dataframe that just about fits into 64GB RAM but persisting stuff to the datastore involves pickling which adds extra memory overhead and gives an out of memory error.

So instead smaller chunks of matches are produced and then persisted to S3 to avoid memory errors.

I should've added a comment why in here, my bad.